### PR TITLE
feat: update `keychain::Balance` to use `bitcoin::Amount`

### LINKED ISSUE #823: `Balance` should use `bitcoin::Amount`
**Describe the bug**  
Currently, the `Balance` type returned by `get_balance()` only uses non-descript `u64`s, while not even stating a denomination anywhere in the documentation (I assume it's sats?). To avoid any confusion and make this much less error prone BDK should reuse the `Amount` type provided by `rust-bitcoin`.

Also, the documentation could generally be improved, e.g., clarifying what exactly it means that some funds are `immature`/`trusted_pending`/`untrusted_pending`/`confirmed`.

**To Reproduce**  
Call `Wallet::get_balance()`.

**Expected behavior**  
Receive a meaningful return value.

**Additional context**  
This is in particular confusing since similar `u64` values are used in the Lightning space very often to denominate millisatoshis. It's best practice to append `_msat` to the names of such variables, which however still doesn't make use of the type system to avoid any conversion errors (we can't, as `Amount` doesn't  allow to express millisatoshis). As BDK doesn't have to deal with millisatoshis, it really should make use of `Amount` here.

### PR DESCRIPTION
**Pull request by oleonardolima**
fixes #823 

### Description

It's being used on `Balance`, and throughout the code, an `u64` represents the amount, which relies on the user to infer its sats, not millisats, or any other representation.

It updates the usage of `u64` on `Balance`, and other APIs:
- `TxParams::add_recipient`
- `KeyChainTxOutIndex::sent_and_received`, `KeyChainTxOutIndex::net_value`
-  `SpkTxOutIndex::sent_and_received`, `SpkTxOutIndex::net_value`

### Notes to the reviewers

It updates some of the APIs to expect the `bitcoin::Amount`, but it does not update internal usage of u64, such as `TxParams` still expects and uses `u64`, please see the PR comments for related discussion.

### Changelog notice

- Changed the `keychain::Balance` struct fields to use `Amount` instead of `u64`.
- Changed the `add_recipient` method on `TxBuilder` implementation to expect `bitcoin::Amount`.
- Changed the `sent_and_received`, and `net_value` methods on `KeyChainTxOutIndex` to expect `bitcoin::Amount`.
- Changed the `sent_and_received`, and `net_value` methods on `SpkTxOutIndex` to expect `bitcoin::Amount`.

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

### CODE REVIEW

File: crates/bdk/src/wallet/tx_builder.rs
Comment (oleonardolima) [inline crates/bdk/src/wallet/tx_builder.rs]: what do you folks think ? do you see any problem in tackling this change as well ?
cc @notmandatory @ValuedMammal @storopoli

File: crates/chain/src/spk_txout_index.rs
Comment (oleonardolima) [inline crates/chain/src/spk_txout_index.rs]: same as above, what do you folks think ? do you see any problem in tackling this change as well ?
cc @notmandatory @ValuedMammal @storopoli

File: crates/bdk/src/wallet/tx_builder.rs
Comment (oleonardolima) [reply crates/bdk/src/wallet/tx_builder.rs]: _NOTE:_ the changes aren't hard (will do on separate commits anyway), it just impact some APIs

Comment (ValuedMammal) [reply crates/bdk/src/wallet/tx_builder.rs]: yeah I'm open to having this change assuming we call `to_sat` before pushing onto recipients.

Comment (oleonardolima) [reply crates/bdk/src/wallet/tx_builder.rs]: hmm, that's ok too, but why we can't we use Amount on the recipients, as well ?

File: crates/chain/src/spk_txout_index.rs
Comment (storopoli) [reply crates/chain/src/spk_txout_index.rs]: I don't see an issue in changing this to `Amount`.

File: crates/bdk/src/wallet/tx_builder.rs
Comment (storopoli) [reply crates/bdk/src/wallet/tx_builder.rs]: Same, go for it.

Comment (ValuedMammal) [reply crates/bdk/src/wallet/tx_builder.rs]: > why we can't we use Amount on the recipients, as well ?

I don't think it adds anything, because the library assumes all amounts are in satoshis and we should keep this invariant.

Comment (notmandatory) [reply crates/bdk/src/wallet/tx_builder.rs]: @ValuedMammal do you mean not to use `Amount` for the `amount` param? I agree internally in `TxParams` it should stay sats but for the public API wouldn't `Amount` be safer?

Comment (notmandatory) [reply crates/bdk/src/wallet/tx_builder.rs]: @oleonardolima yes thanks for taking this one on, it's one we'd still like to get into the 1.0 alpha.

Comment (ValuedMammal) [reply crates/bdk/src/wallet/tx_builder.rs]: yes for changing this method `add_recipient` to accept an `Amount`. I don't think the type in `TxParams` needs to change in this PR

Comment (oleonardolima) [reply crates/bdk/src/wallet/tx_builder.rs]: Good, I think we should update the `set_recipient` to accept `Amount` too, right ? @ValuedMammal @notmandatory wdyt ?

**Comment by oleonardolima** [comment]
I think it's ready for a new round of reviews

_I'll move the previous conversations to resolved after it, and if it's everything looks good on this new review round)_

**Comment by storopoli** [comment]
> I think it's ready for a new round of reviews

Can you move it from Draft PR to a full PR?

**Comment by oleonardolima** [comment]
> > I think it's ready for a new round of reviews
> 
> Can you move it from Draft PR to a full PR?

Yes, forgot to do it

**Comment by oleonardolima** [comment]
~Oh, I forgot to run the tests on docs, working on it 😅~
Done!

File: crates/bdk/src/wallet/tx_builder.rs
Comment (notmandatory) [reply crates/bdk/src/wallet/tx_builder.rs]: Yes anywhere in the public API where we can replace a `u32` (or `i64`) representing sats with `Amount` (or `SignedAmount`) I'd like to do it it.

File: crates/bdk/src/wallet/coin_selection.rs
Comment (storopoli) [inline crates/bdk/src/wallet/coin_selection.rs]: Isn't this covered in the

```rust
//! # use bitcoin::*;
```

on line 28?

```suggestion
//!     builder.add_recipient(
//!         to_address.script_pubkey(),
//!         Amount::from_sat(50_000),
//!     );
```

File: crates/bdk/src/wallet/mod.rs
Comment (storopoli) [inline crates/bdk/src/wallet/mod.rs]: It's already imported in line 40

```suggestion
    ///        .add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000));
```

File: crates/bdk/src/wallet/mod.rs
Comment (storopoli) [inline crates/bdk/src/wallet/mod.rs]: Same

```suggestion
    ///         .add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000))
```

File: crates/bdk/src/wallet/mod.rs
Comment (storopoli) [inline crates/bdk/src/wallet/mod.rs]: Same

```suggestion
    ///     builder.add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000));
```

File: crates/bdk/src/wallet/tx_builder.rs
Comment (storopoli) [inline crates/bdk/src/wallet/tx_builder.rs]: Already imported in line 18

```suggestion
//!     .add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000))
```

File: crates/bdk/src/wallet/tx_builder.rs
Comment (storopoli) [inline crates/bdk/src/wallet/tx_builder.rs]: You are importing in L50

```suggestion
///         .add_recipient(addr1.script_pubkey(), Amount::from_sat(50_000))
///         .add_recipient(addr2.script_pubkey(), Amount::from_sat(50_000));
```

File: crates/bdk/src/wallet/tx_builder.rs
Comment (storopoli) [inline crates/bdk/src/wallet/tx_builder.rs]: imported L50

```suggestion
///         builder.add_recipient(addr.script_pubkey(), Amount::from_sat(50_000));
```

File: crates/bdk/src/wallet/tx_builder.rs
Comment (storopoli) [inline crates/bdk/src/wallet/tx_builder.rs]: imported L50

```suggestion
    ///         Amount::from_sat(50_000),
```

File: crates/bdk/tests/psbt.rs
Comment (storopoli) [inline crates/bdk/tests/psbt.rs]: Why this change?

File: crates/chain/src/keychain.rs
Comment (storopoli) [inline crates/chain/src/keychain.rs]: Why not import `bitcoin::Amount` and reduce the clutter?

File: crates/chain/src/keychain/txout_index.rs
Comment (storopoli) [inline crates/chain/src/keychain/txout_index.rs:292]: Why this needs to be a `SignedAmount`?

File: crates/chain/src/spk_txout_index.rs
Comment (storopoli) [inline crates/chain/src/spk_txout_index.rs:308]: Why this needs to be a `SignedAmount`?

File: crates/chain/src/tx_graph.rs
Comment (storopoli) [inline crates/chain/src/tx_graph.rs]: Why not import `bitcoin::Amount` and reduce the clutter?

File: crates/chain/tests/test_spk_txout_index.rs
Comment (storopoli) [inline crates/chain/tests/test_spk_txout_index.rs:37]: Why this needs to be a `SignedAmount`?

File: crates/chain/tests/test_spk_txout_index.rs
Comment (storopoli) [inline crates/chain/tests/test_spk_txout_index.rs:79]: Why this needs to be a `SignedAmount`?

**Review by storopoli — CHANGES_REQUESTED** [review]
Some suggestions.

Most important, why we are using `SignedAmount` in some places?

File: crates/chain/src/keychain/txout_index.rs
Comment (oleonardolima) [reply crates/chain/src/keychain/txout_index.rs:292]: It's because `net_value` used `i64` previously, and to keep the standard (and makes sense in my opinion given the  fn name) I used the `SignedAmount`, same for the other comments, they are all `net_value` references.

File: crates/bdk/tests/psbt.rs
Comment (oleonardolima) [reply crates/bdk/tests/psbt.rs]: I have no idea, probably rust-analyzer stuff 🤔, fixed it.

File: crates/chain/src/keychain.rs
Comment (oleonardolima) [reply crates/chain/src/keychain.rs]: Done, much better now :)

**Comment by oleonardolima** [comment]
> Some suggestions.
> 
> Most important, why we are using `SignedAmount` in some places?

It's because net_value used i64 previously, and to keep the standard (and makes sense in my opinion given the fn name) I used the SignedAmount, same for the other comments, they are all net_value references.

_(replying on the overall comment too)_

File: crates/chain/src/keychain/txout_index.rs
Comment (storopoli) [reply crates/chain/src/keychain/txout_index.rs:292]: Oh I see...

**Review by storopoli — APPROVED** [review]
ACK 362a67aefd5baea6a321721df05e9aff79894c86

Great work!

File: crates/chain/src/spk_txout_index.rs
Comment (ValuedMammal) [inline crates/chain/src/spk_txout_index.rs]: I think you want to do `to_signed` on each Amount before doing the subtraction.

Comment (oleonardolima) [reply crates/chain/src/spk_txout_index.rs]: Yes, you're right, fixed it.

File: example-crates/example_cli/src/lib.rs
Comment (ValuedMammal) [inline example-crates/example_cli/src/lib.rs:509]: This function intends to print out a value in sats, so for this to work we need another `to_sat` inside the function.

```diff
             fn print_balances<'a>(
                 title_str: &'a str,
                 items: impl IntoIterator<Item = (&'a str, Amount)>,
             ) {
                 println!("{}:", title_str);
                 for (name, amount) in items.into_iter() {
-                    println!("    {:<10} {:>12} sats", name, amount)
+                    println!("    {:<10} {:>12} sats", name, amount.to_sat())
```

File: crates/chain/src/spk_txout_index.rs
Comment (ValuedMammal) [inline crates/chain/src/spk_txout_index.rs]:
```suggestion
        received.to_signed().expect("valid SignedAmount")
            - sent.to_signed().expect("valid SignedAmount")
```

File: example-crates/example_cli/src/lib.rs
Comment (oleonardolima) [reply example-crates/example_cli/src/lib.rs:509]: Nice, good catch! Fixed it.

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs:981]: In this case, the docs for this method shouldn't say that the returned amounts are "in satoshis".

File: crates/bdk/tests/wallet.rs
Comment (ValuedMammal) [inline crates/bdk/tests/wallet.rs]: If you mean can we change `Wallet::calculate_fee` to return an `Amount`, I think possibly yes. But to avoid scope creep we should push it to a future PR.

```suggestion
```

**Review by ValuedMammal — COMMENTED** [review]
Sorry my reviews have been intermittent. I think this is good.

The commit history could be cleaned up slightly:

- In df92624b011668216bba9908f6bd4fbb6db06457 we want to refer to `TxBuilder::add_recipient` (not CoinSelectionAlgorithm)

- Some changes in later commits (like fixing imports and wallet examples) could be squashed into earlier ones, but this not a blocker in my opinion.

Some other thoughts: I agree there is potentially more to do like updating `calculate_fee` and `Psbt::fee_amount` to use `Amount`. I'm still unsure how we should handle amounts internally. Honestly I don't mind leaving it as `u64`. If we decide that all amounts should be strongly typed, then `Amount` seems like a logical choice, my only concern is we should avoid the possibility of mixing up denominations.

File: crates/bdk/tests/wallet.rs
Comment (oleonardolima) [reply crates/bdk/tests/wallet.rs]: Yes, that's what I mean. Cool, I'll address it in another PR.

File: crates/bdk/src/wallet/mod.rs
Comment (oleonardolima) [reply crates/bdk/src/wallet/mod.rs:981]: Yes, updated it, thanks!

**Comment by oleonardolima** [comment]
> Sorry my reviews have been intermittent. I think this is good.
> 
> The commit history could be cleaned up slightly:
> 
> * In [df92624](https://github.com/bitcoindevkit/bdk/commit/df92624b011668216bba9908f6bd4fbb6db06457) we want to refer to `TxBuilder::add_recipient` (not CoinSelectionAlgorithm)
> * Some changes in later commits (like fixing imports and wallet examples) could be squashed into earlier ones, but this not a blocker in my opinion.
> 
> Some other thoughts: I agree there is potentially more to do like updating `calculate_fee` and `Psbt::fee_amount` to use `Amount`. I'm still unsure how we should handle amounts internally. Honestly I don't mind leaving it as `u64`. If we decide that all amounts should be strongly typed, then `Amount` seems like a logical choice, my only concern is we should avoid the possibility of mixing up denominations.

@ValuedMammal No worries, thanks for the comments and suggestions! I tried to address them, please let me know if something remains or if new needed changes appear.

File: crates/bdk/src/wallet/mod.rs
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs]: What do you think of this?

```suggestion
    /// Compute the `tx`'s sent and received [`Amount`]s.
```

**Review by ValuedMammal — APPROVED** [review]
ACK 2e6aeb7499d9c4fadce4024fe26666976447ac2e

**Comment by ValuedMammal** [comment]
Actually, changing `calculate_fee` and `Psbt::fee_amount` should be fairly easy to add to this PR but I'll let you decide if you want to tackle it here or later. Besides that, I'm not sure if there's still more places in the public API that would benefit from using `Amount`.

**Comment by oleonardolima** [comment]
> Actually, changing `calculate_fee` and `Psbt::fee_amount` should be fairly easy to add to this PR but I'll let you decide if you want to tackle it here or later. Besides that, I'm not sure if there's still more places in the public API that would benefit from using `Amount`.

I updated `calculate_fee` and `fee_amount` on ac762f4a862db227432aa7daeeee424d2b653af3, but ended up touching other parts as well.

I would rather move this commit ac762f4a862db227432aa7daeeee424d2b653af3 to another PR, and keep only the previous ones here, wdyt ?

**Comment by oleonardolima** [comment]
> > Actually, changing `calculate_fee` and `Psbt::fee_amount` should be fairly easy to add to this PR but I'll let you decide if you want to tackle it here or later. Besides that, I'm not sure if there's still more places in the public API that would benefit from using `Amount`.
> 
> I updated `calculate_fee` and `fee_amount` on [ac762f4](https://github.com/bitcoindevkit/bdk/commit/ac762f4a862db227432aa7daeeee424d2b653af3), but ended up touching other parts as well.
> 
> I would rather move this commit [ac762f4](https://github.com/bitcoindevkit/bdk/commit/ac762f4a862db227432aa7daeeee424d2b653af3) to another PR, and keep only the previous ones here, wdyt ?

I moved it to the other PR #1426, we can start discussing the changes on that one after this one is merged.

**Review by evanlinjin — APPROVED** [review]
ACK 22aa534d7648e5808414ea3adfcfb702572bd6c9

**Pull request merged**
